base/Base/Base_Base_implementation.F90

```diff
@@ -3169,6 +3169,174 @@ end function ijsearch
+       ! edge of face 1 along longitude
+       if ( I1 ==1 .and. J2<=IM_WORLD ) then
+          allocate(corner_lons(I2-I1+2, J2-J1+2))
+          allocate(corner_lats(I2-I1+2, J2-J1+2))
+          call MAPL_GridGetCorners(Grid,corner_lons,corner_lats,rc=status)
+          if ( J1 == 1) then
+             !print*, "acurate corner lat lon: ", -alpha,  1.750d0*MAPL_PI_R8 - Japan_shift
+             !print*, "from getCorner lat lon: ", corner_lats(1,1), corner_lons(1,1)
+             _ASSERT( abs((1.750d0*MAPL_PI_R8 - Japan_shift) - corner_lons(1,1)) <= epsilon(1.0), "Grid should have pi/18 shift")
+          endif
+          do j = J1, J2+1
+             c_lat = -alpha + (j-1)*dalpha 
+             _ASSERT( abs(c_lat - corner_lats(1,j-J1+1)) <= 10.0*epsilon(1.0), "could be 1)Grid is NOT gnomonic_ed, 2)lats lons from MAPL_GridGetCorners are NOT accurate (single precision from ESMF)")
+             !if ( abs(c_lat - corner_lats(1,j-J1+1)) > epsilon(1.0)) then
+             !  print*, "Grid should be gnomonic_ed: ", c_lat, corner_lats(1,j-J1+1) , j
+             !endif
+          enddo
+          deallocate(corner_lons, corner_lats)
+       endif
+       grid_is_ok = .true.
+    endif
+
+    allocate(lons(npts),lats(npts))
+    if (present(lon) .and. present(lat)) then
+       lons = lon
+       lats = lat
+    else if (present(lonR8) .and. present(latR8)) then
+       lons = lonR8
+       lats = latR8
+    end if
+
+    !shift the conner away from Japan Fuji Mt.
+    lons = lons + Japan_shift
+
+    ! get xyz from sphere surface
+    allocate(xyz(3, npts), max_abs(npts))
+    xyz(1,:) = cos(lats)*cos(lons)
+    xyz(2,:) = cos(lats)*sin(lons)
+    xyz(3,:) = sin(lats)
+
+    ! project onto cube
+    max_abs = maxval(abs(xyz), dim=1)
+
+    xyz(1,:) = xyz(1,:)/max_abs
+    xyz(2,:) = xyz(2,:)/max_abs
+    xyz(3,:) = xyz(3,:)/max_abs
+
+    allocate(x, source=xyz(1,:))
+    allocate(y, source=xyz(2,:))
+    allocate(z, source=xyz(3,:))
+
+    II = -1
+    JJ = -1
+
+    ! Here the error and shift are used for edge points.
+    ! the edge points are assigned in the order of face 1,2,3,4,5,6
+    error = epsilon(1.0d0)
+    shift = 10.0d0*error
+
+    ! face = 1
+    where ( abs(x-1.0d0) <= error )
+       ! Four edges assigned to face 1
+       where (abs(y - 1.0d0) <=error) y = y-shift
+       where (abs(y + 1.0d0) <=error) y = y+shift
+       where (abs(z - 1.0d0) <=error) z = z-shift
+       where (abs(z + 1.0d0) <=error) z = z+shift
+       II = ceiling((atan(y/sqr2) + alpha)/dalpha)
+       JJ = ceiling((atan(z/sqr2) + alpha)/dalpha)
+    endwhere
+
+    ! face = 2
+    where (abs(y-1.0d0) <= error)
+       ! Three edges assigned to face 2
+       where (abs(x + 1.0d0) <=error) x = x+shift
+       where (abs(z - 1.0d0) <=error) z = z-shift
+       where (abs(z + 1.0d0) <=error) z = z+shift
+       II = ceiling((atan(-x/sqr2) + alpha)/dalpha)
+       JJ = ceiling((atan(z/sqr2) + alpha)/dalpha)
+       JJ = JJ + IM_WORLD
```

Comment:
I think it is probably good to make this expression a function of x and z.   You would then just pass different values in each of the blocks for the 6 faces.    "FACE_ID" would be one of the parameters used to do the fake JJ

Reply:
Cannot use function inside a where struct